### PR TITLE
Fixing address re-use and countdown timer

### PR DESCRIPTION
**Pull request by cnohall**
This is similar to the previous PR (1.7.84). However, this time only issues #92 (Address Reuse is not happening) and #93(Refresh Payment Page Timer) are addressed, to keep it nice and simple. Even though this PR says 117 additions and 89 deletions, it's mostly formatting/slight change of location that marks previous code as deleted and then added. I'll create a separate PR for the Issue #91 (Minor Issues) and one PR for issue #89 (Some Translation strings are not correct [PS 1.7]).

### CODE REVIEW

File: controllers/front/validation.php [hunk 1]
```diff
@@ -119,101 +111,128 @@ public function postProcess()
...
-                    (int) $product['id_product_attribute']
-                );
-        }
-        if ($this->context->cookie->id_guest) {
-            $guest = new Guest($this->context->cookie->id_guest);
-            $new_cart->mobile_theme = $guest->mobile_theme;
-        }
+        $sql = 'SELECT * FROM '._DB_PREFIX_."blockonomics_bitcoin_orders WHERE id_cart = $cart->id";
+        $order = Db::getInstance()->getRow($sql);
+        $time_left = $this->get_time_left($order);
```
Comment (DarrenWestwood) [inline controllers/front/validation.php:116]: Let's use `time_remaining` instead of `time_left` to maintain consistency between platforms.

**Review by DarrenWestwood — CHANGES_REQUESTED** [review]
Address reuse and the payment timer are working well and follow the design.

File: controllers/front/validation.php (hunk 1 above)
Comment (cnohall) [reply controllers/front/validation.php]: Done.

File: controllers/front/validation.php [hunk 2]
```diff
@@ -119,101 +111,128 @@ public function postProcess()
...
+                    (int) $bits .
+                    "', 0)"
+            );
+        } else {
+            $address = $order['addr'];
+            $id_order = $order['id_order'];
+            $current_time = $order['timestamp'];
+            //if value of cart has changed because the user has added or removed products from the cart
+            if (!$time_left) {
+                $query = "UPDATE "._DB_PREFIX_."blockonomics_bitcoin_orders SET timestamp=".time().", value=$total, bits=$bits WHERE id_cart = $cart->id";
```
[the remark this answers is not in the document]
Comment (cnohall) [reply controllers/front/validation.php:213]: True, made the necessary changes and pushed it.

File: controllers/front/validation.php [hunk 3]
```diff
@@ -98,14 +98,6 @@ public function postProcess()
...
 
-        $responseObj = $blockonomics->getNewAddress();
-
-        if (!$responseObj->data || !$responseObj->data->address) {
-            $this->displayError($blockonomics);
-        }
-
-        $new_address = $responseObj->data->address;
-
         $current_time = time();
```
[the remark this answers is not in the document]
Comment (cnohall) [reply controllers/front/validation.php:101]: Done.

**Comment by cnohall** [comment]
I've updated the PR according to @DarrenWestwood's requests.

File: controllers/front/validation.php [hunk 4]
```diff
@@ -98,122 +98,135 @@ public function postProcess()
...
-        }
-        if ($this->context->cookie->id_guest) {
-            $guest = new Guest($this->context->cookie->id_guest);
-            $new_cart->mobile_theme = $guest->mobile_theme;
-        }
+            // Validate the order
+            $mailVars = array(
+                '{bitcoin_address}' => $address,
+                '{bits}' => $bits / 1.0e8,
+                '{track_url}' =>
```
Comment (DarrenWestwood) [inline controllers/front/validation.php:143]: track_url should be removed during https://github.com/blockonomics/prestashop-plugin/pull/90. Please confirm this is not required and remove it accordingly.

File: controllers/front/validation.php [hunk 5]
```diff
@@ -98,122 +98,135 @@ public function postProcess()
...
+                    "','" .
+                    (int) $bits .
+                    "', 0)"
+            );
+        } else {
+            $address = $order['addr'];
+            $id_order = $order['id_order'];
+            $current_time = $order['timestamp'];
+            $time_remaining = $this->get_time_remaining($order);
+            //if value of cart has changed because the user has added or removed products from the cart
```
Comment (DarrenWestwood) [inline controllers/front/validation.php:204]: This comment is incorrect and can be moved/removed.

**Review by DarrenWestwood — APPROVED** [review]
The address and timer reuse works as desired and the code looks good. Let's check the changes through the PS module validator and fix any issues.

File: controllers/front/validation.php [hunk 6]
```diff
@@ -225,9 +238,7 @@ public function postProcess()
             'currency_iso_code' => $currency->iso_code,
             'bits_payed' => 0,
             'redirect_link' => $redirect_link,
-            'timeperiod' => Configuration::get(
-                'BLOCKONOMICS_TIMEPERIOD'
-            )
+            'timeperiod' => $time_remaining
```
Comment (DarrenWestwood) [inline controllers/front/validation.php:241]: The timeperiod is used as the total time in the progress bar. Setting $time_remaining to timeperiod causes the progress bar to reset on refreshes incorrectly:
PrestaShop3

The time remaining should instead be passed as a new key - `'time_remaining' => $time_remaining` and timeperiod reinstated.

File: controllers/front/validation.php (hunk 5 above)
Comment (cnohall) [reply controllers/front/validation.php:204]: True, the comment has now been removed.

File: controllers/front/validation.php (hunk 6 above)
Comment (cnohall) [reply controllers/front/validation.php:241]: Done, now the progressbar works at it should.

File: controllers/front/validation.php (hunk 4 above)
Comment (cnohall) [reply controllers/front/validation.php:143]: I've tested creating a new order after removing `track_url`. The checkout works fine without it. However, it looks like `track_url` is a part of the email that should go out to the customer after placing an order? Is there a way of testing or looking on an outgoing email with and without `track_url`?

This branch is based on the 1.7 branch after it was merged with #90. There should be no merge conflicts after removing `track_url`. Just to want to make sure that the outgoing email still looks like it should without `track_url`.

Comment (DarrenWestwood) [reply controllers/front/validation.php:143]: Test outgoing mails from Prestashop by setting up your mail server/smtp info from the Prestashop Admin > Advanced Parameters > Email.
Order confirmation emails use the order_conf.html template which does not include track_url.

Comment (cnohall) [reply controllers/front/validation.php]: Thanks, now I've tested the email after removing `track_url` and it works, just as you said. Just feels better to give it a go to make sure.

**Comment by cnohall** [comment]
Updated the PR according to the comments from Darren's review.

**Comment by DarrenWestwood** [comment]
The changes are working as desired. Please update the changes according to the Prestashop Validator. There are some Optimizations, Licenses, and Standards that should be resolved.

**Comment by cnohall** [comment]
Seems like I need a contributor's/seller account to be able to use Prestashop Validator. Is there a way around it?

**Comment by cnohall** [comment]
Now it's brought up to speed, the same as #98 (minor-issues) PR. Only licence date and folder structure that remains as an error. However, it's probably best to fix that after everything is merged into 1.7.8.4, so to not create any conflicts.

File: controllers/front/validation.php [hunk 7]
```diff
@@ -237,6 +238,28 @@ public function postProcess()
...
+            $time_remaining = ($order['timestamp'] +
+            (Configuration::get('BLOCKONOMICS_TIMEPERIOD') * 60) - time()) / 60;
+            if ($time_remaining > 0) {
+                return $time_remaining;
+            }
+        }
+        return false;
+    }
+
+    private function getBits($blockonomics, $currency, $total)
```
Comment (DarrenWestwood) [inline controllers/front/validation.php:253]: get_bits has been renamed and requires fixing:
127-0-0-1-prestashop3-index-php-fc-module-module-blockonomics-controller-validation-id_lang-2

Comment (cnohall) [reply controllers/front/validation.php:253]: Yes. Now I've corrected both get_bits and get_remaining_time. Both when defining and calling them.

**Review by DarrenWestwood — APPROVED** [review]
PR approved and ready for merge

**Pull request merged**